src/vptranslator.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2017 Kisio Digital and/or its affiliates.
+//
+// This program is free software: you can redistribute it and/or
+// modify it under the terms of the GNU General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see
+// <http://www.gnu.org/licenses/>.
+
+//use crate::common_format;
+
+use chrono::{Datelike, Duration, NaiveDate, Weekday};
+use num_traits::FromPrimitive;
+use std::collections::BTreeSet;
+use std::vec::Vec;
+
+#[derive(Debug, PartialOrd)]
+struct ValidityPeriod {
+    start_date: NaiveDate,
+    end_date: NaiveDate,
+}
+
+impl Default for ValidityPeriod {
+    fn default() -> Self {
+        ValidityPeriod {
+            start_date: chrono::naive::MIN_DATE,
+            end_date: chrono::naive::MIN_DATE,
+        }
+    }
+}
+
+impl PartialEq for ValidityPeriod {
+    fn eq(&self, other: &Self) -> bool {
+        (self.start_date, &self.end_date) == (other.start_date, &other.end_date)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum ExceptionType {
+    Sub,
+    Add,
+}
+
+#[derive(Debug)]
+struct ExceptionDate {
+    date: NaiveDate,
+    exception_type: ExceptionType,
+}
+
+impl PartialEq for ExceptionDate {
+    fn eq(&self, other: &ExceptionDate) -> bool {
+        self.exception_type == other.exception_type
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct BlockPattern {
+    week: u8,
+    nb_weeks: i64,
+    operating_days: Vec<Weekday>,
+    validity_periods: Vec<ValidityPeriod>,
+    exceptions: Vec<ExceptionDate>,
+}
+
+fn get_prev_monday(date: NaiveDate) -> NaiveDate {
+    let res = date + Duration::days(-1 * date.weekday().num_days_from_monday() as i64);
+    res
+}
+
+fn weeks(dates: &BTreeSet<NaiveDate>) -> Vec<u8> {
+    let start_date: NaiveDate = get_prev_monday(*dates.iter().next().unwrap());
```

Comment:
each `unwrap` is a kitty dying somewhere :laughing: 

maybe you can rewrite it like:

```rust
    let start_date = match dates.iter().next().map(|d|get_prev_monday(*d)) {
        Some(d) => d,
        None => return vec![]
    };
```